mithril-aggregator/src/services/prover.rs

```diff
@@ -70,25 +70,37 @@ impl MithrilProverService {
         &self,
         transactions: Vec<CardanoTransaction>,
     ) -> StdResult<MKMap<BlockRange, MKMapNode<BlockRange>>> {
-        let mut transactions_by_block_ranges: HashMap<BlockRange, Vec<TransactionHash>> =
-            HashMap::new();
+        let mut transactions_by_block_ranges: BTreeMap<BlockRange, Vec<TransactionHash>> =
+            BTreeMap::new();
+        let mut last_transaction: Option<CardanoTransaction> = None;
```

Comment:
I think we could maybe avoid cloning the transactions using `last_transaction_block_number: Option<u64>` instead of  `last_transaction: Option<CardanoTransaction>`.

```
    fn compute_merkle_map_from_transactions(
        &self,
        transactions: Vec<CardanoTransaction>,
    ) -> StdResult<MKMap<BlockRange, MKMapNode<BlockRange>>> {
        let mut transactions_by_block_ranges: BTreeMap<BlockRange, Vec<TransactionHash>> =
            BTreeMap::new();
        let mut last_transaction_block_number: Option<u64> = None;
        for transaction in transactions {
            let block_range = BlockRange::from_block_number(transaction.block_number);
            last_transaction_block_number = Some(transaction.block_number);
            transactions_by_block_ranges
                .entry(block_range)
                .or_default()
                .push(transaction.transaction_hash);
        }
        let mut block_ranges = transactions_by_block_ranges.into_iter().try_fold(
            vec![],
            |mut acc, (block_range, transactions)| -> StdResult<Vec<(_, MKMapNode<_>)>> {
                acc.push((block_range, MKTree::new(&transactions)?.into()));
                Ok(acc)
            },
        )?;

        // This is a temporary fix to avoid including an incomplete block ranges in the computation of the prover.
        // This will be swiftly replaced by the new way of computing proof relying on the block range roots stored in database.
        if let Some(block_number) = last_transaction_block_number {
            if let Some((last_block_range, _)) = block_ranges.last() {
                if block_number < last_block_range.end - 1 {
                    block_ranges.pop();
                }
            }
        }

        let mk_hash_map = MKMap::new_from_iter(
            block_ranges
        )
        .with_context(|| "ProverService failed to compute the merkelized structure that proves ownership of the transaction")?;

        Ok(mk_hash_map)
    }
```

Reply:
FYI, this function will be removed with [#1635](https://github.com/input-output-hk/mithril/issues/1635)

Reply:
Done